creator-node/src/services/stateMachineManager/stateReconciliation/recoverOrphanedData.jobProcessor.ts

```diff
@@ -18,40 +31,178 @@ const { QUEUE_NAMES } = require('../stateMachineConstants')
+    WALLETS_ON_NODE_KEY,
+    walletsOnThisNodeArr
+  )
+  return numWalletsOnNode
+}
+
+/**
+ * Queries the given discovery node to find all users who have this content node as their primary or secondary.
+ * Adds them to a redis set.
+ */
+const _storeWalletsWithThisNodeInReplica = async (
+  discoveryNodeEndpoint: string
+) => {
+  await redisClient.del(WALLETS_WITH_NODE_IN_REPLICA_SET_KEY)
+
+  // Make paginated Discovery queries to find all wallets with this current node in their replica set (primary or secondary)
+  let prevUserId = 0
+  let batchOfUsers: StateMonitoringUser[] = []
+  const walletsWithNodeInReplicaSetArr: string[] = []
+  do {
+    // Get batch of users and all it to the array of all wallets
+    batchOfUsers = await getNodeUsers(
+      discoveryNodeEndpoint,
+      thisContentNodeEndpoint,
+      prevUserId,
+      NUM_USERS_PER_QUERY
+    )
+    batchOfUsers?.forEach((user) =>
+      walletsWithNodeInReplicaSetArr.push(user.wallet)
+    )
+
+    // Move pagination cursor to the end of the batch
+    prevUserId = batchOfUsers?.length
+      ? batchOfUsers[batchOfUsers.length - 1].user_id
+      : 0
+  } while (batchOfUsers?.length === NUM_USERS_PER_QUERY && prevUserId !== 0)
+
+  // Save the wallets as a redis set and return the set cardinality
+  const numWalletsWithNodeInReplicaSet = await redisClient.sadd(
+    WALLETS_WITH_NODE_IN_REPLICA_SET_KEY,
+    walletsWithNodeInReplicaSetArr
+  )
+  return numWalletsWithNodeInReplicaSet
+}
+
+/**
+ * Gets a sync request that can be issued to "force wipe" (modified forceResync=true flag):
+ *  1. Merges orphaned data from this node into the user's primary.
+ *  2. Wipes the user's data from this node.
+ *  3. *DON'T* resync from the primary to this node -- this is what non-modified forceResync would do.
+ * @param wallet the wallet of the user with data orphaned on this node
+ */
+const _getSyncReqToRecoverOrphanedData = async (
+  wallet: string,
+  logger: Logger
+): Promise<IssueSyncRequestJobParams | undefined> => {
+  const primaryEndpoint = await _getPrimaryForWallet(wallet, logger)
+  if (!primaryEndpoint) return undefined
+
+  try {
+    const { syncReqToEnqueue } = getNewOrExistingSyncReq({
+      userWallet: wallet,
+      primaryEndpoint,
+      secondaryEndpoint: thisContentNodeEndpoint,
+      syncType: SyncType.Recurring,
+      syncMode: SYNC_MODES.MergePrimaryThenWipeSecondary
+    })
+
+    return syncReqToEnqueue
+  } catch (e: any) {
+    logger.error(
+      `Error getting sync request to recover orphaned data for user ${wallet} with primary ${primaryEndpoint} - ${e.message}`
+    )
+    return undefined
   }
 }
+
+const _getPrimaryForWallet = async (
+  wallet: string,
+  logger: Logger
+): Promise<string> => {
+  return 'primaryEndpoint'
+  // TODO: The issue here is that there's no route to get the user's RS from their wallet.
+  //       The `users` table on Discovery has creator_node_endpoint+wallet, so we could make a route if needed.
```

Comment:
this one is a bit tricky but maybe there's a simple way I'm missing that doesn't involve adding a new route @SidSethi @dmanjunath ? like some other table join to get user_id and make a Discovery request to get primary endpoint using user_id?

Reply:
we do this all over the place, no? like in `middlewares` for `ensurePrimaryMiddleware`
also discovery getUsers will return user replica set https://discoveryprovider2.staging.audius.co/users?wallet=0x280a556fa6c04341e87829248ae517e4047be681

we should likely do this from chain though, to avoid indexing discovery

this card we just talked about is actually pretty relevant https://linear.app/audius/issue/CON-332/improved-cn-endpointspid-retrieval-logic
this data already exists in monitor-state job processor, no?

lets chat about this on call briefly, along with other concerns

Reply:
could've sworn I tried the `users?wallet=` queryparam 🤦 this should be simple then, thanks!